File: 383-ransom-note/ransom-note.go
import "fmt"
func canConstruct(ransomNote string, magazine string) bool {
	charCount := make(map[rune]int)
    
    for _, char := range magazine {
        charCount[char]++
    }

    for _, char := range ransomNote {
        if charCount[char] == 0 {
            return false
        }
        charCount[char]--
    }

    return true
}

File: 383-ransom-note/ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        Map<Character, Integer> countMap = new HashMap<>();
        for (char c:magazine.toCharArray()) {
            countMap.put(c, countMap.getOrDefault(c, 0)+1);
        }


        for (char c:ransomNote.toCharArray()) {
            if (countMap.getOrDefault(c, 0)==0){
                return false;
            }
            countMap.put(c, countMap.getOrDefault(c, 0)-1);
        }

        return true;
    }
}
